### ISSUE TITLE
How should the program be modified when the feature distribution of the source domain is inconsistent with the source domain data distribution?"

### ISSUE THREAD

**Issue by cyy521xx** [mshunya/AffineTL#1]
Hello author, I have carefully studied your code and found that in the three experiments, the features of the source domain and the target domain are the same. I want to implement transfer learning when the features of the source domain and the target domain are inconsistent. May I ask how I should modify the program? Looking forward to your reply, thank you."

**Comment by mshunya** [comment]
In some of our experiments, we used different feature sets for the source and target data.

For example, in the notebook `200_TransferLearning.ipynb` located in the repository named `12_LTC`, the class `cls_InvTrans_NN` takes concatenated features `concatenate(X_source, X_target)` as an input to the fit method and internally splits them as follows:

```python
def fit(self, X, y=None):
    self.X_train = X.iloc[:, :dim_x]
    self.X_source = X.iloc[:, dim_x:]
```

Note that X_train and X_source do not need to have the same feature sets. Indeed, in this experiment, we used different feature sets: intermediate neural network layer features (`ll_train_scal`, 50-100 dimensions) as the source features, and scaled input features (`x_train_scal`, 290 dimensions) as the target features. These were concatenated using:
```python
x_train_adds = pd.merge(x_train_scal, ll_train_scal, left_index=True, right_index=True)
```

Similarly, in the repository `13_Cp`, the experiments involve completely different feature sets: a single-dimensional feature for the source and a 190-dimensional force-field feature for the target. This experiment uses a linear model implementing Algorithm S.1 described in our paper.

If you have any further questions or need additional clarification, please feel free to ask!

**Comment by cyy521xx** [comment]
Hello, I would like to ask for advice on how to modify the code in this scenario: the X data from the source domain and the X data from the target domain are inconsistent. For example, I have a lot of source domain data representing the hardness of alloys (e.g., x1, x2, ... xm, y1, ... ym, where xi are alloy compositions or other features, and yi are hardness values), and only a small amount of data on tensile strength (u1, u2, ... un, v1, v2, ... vn, where m &gt; n, and ui are alloy compositions or other features, and vi are tensile strength values). How can I transfer hardness to tensile strength? I don't know how to modify the code.


------------------ 原始邮件 ------------------
发件人:                                                                                                                        "mshunya/AffineTL"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年4月27日(星期天) 下午4:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [mshunya/AffineTL] How should the program be modified when the feature distribution of the source domain is inconsistent with the source domain data distribution?" (Issue #1)



 mshunya left a comment (mshunya/AffineTL#1)
 
In some of our experiments, we used different feature sets for the source and target data.
 
For example, in the notebook 200_TransferLearning.ipynb located in the repository named 12_LTC, the class cls_InvTrans_NN takes concatenated features concatenate(X_source, X_target) as an input to the fit method and internally splits them as follows:
 def fit(self, X, y=None):     self.X_train = X.iloc[:, :dim_x]     self.X_source = X.iloc[:, dim_x:]
 
Note that X_train and X_source do not need to have the same feature sets. Indeed, in this experiment, we used different feature sets: intermediate neural network layer features (ll_train_scal, 50-100 dimensions) as the source features, and scaled input features (x_train_scal, 290 dimensions) as the target features. These were concatenated using:
 x_train_adds = pd.merge(x_train_scal, ll_train_scal, left_index=True, right_index=True)
 
Similarly, in the repository 13_Cp, the experiments involve completely different feature sets: a single-dimensional feature for the source and a 190-dimensional force-field feature for the target. This experiment uses a linear model implementing Algorithm S.1 described in our paper.
 
If you have any further questions or need additional clarification, please feel free to ask!

**Comment by mshunya** [comment]
Based on your description, my understanding is that your scenario involves completely separate datasets for the source and target domains, such as a source dataset {(x_i, y_i)}{i=1}^m and a target dataset {(u_j, v_j)}{j=1}^n.

Our method assumes that for each sample in the target domain, some relevant information from the source domain is available. In other words, the dataset should be structured as {(u_j, v_j, s_j)}_{j=1}^n, where s_j represents source domain information related to sample j (for instance, hardness or related information).

To apply our method effectively in your scenario, features x and u generally must belong to the same feature space (e.g., both x and u are alloy composition feature vectors). In this case, you would train a predictive model f(x) on the source dataset {(x, y)} and then generate the source-derived features s for the target data by applying f or extracting intermediate features from f. You would then form a dataset {(u, v, s)} to apply our method.

If x and u exist in different feature spaces (for example, x representing alloy composition and u representing structural features), our method cannot be directly applied. In such cases, additional steps like introducing an encoder to map u to the feature space of x would be required.

In short, for our method to be applicable, you need a target dataset of the form {(u_j, v_j, s_j)}_{j=1}^n, where each sample includes (target input, target output, and additional source-derived information).

Please let me know if anything needs further clarification.

**Comment by cyy521xx** [comment]
Thank you very much for your reply, which has helped clear up my confusion. I appreciate it!




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2025年4月27日(星期天) 下午5:10
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [mshunya/AffineTL] How should the program be modified when the feature distribution of the source domain is inconsistent with the source domain data distribution?&quot; (Issue #1)












mshunya left a comment (mshunya/AffineTL#1)
 
Based on your description, my understanding is that your scenario involves completely separate datasets for the source and target domains, such as a source dataset {(x_i, y_i)}{i=1}^m and a target dataset {(u_j, v_j)}{j=1}^n.
 
Our method assumes that for each sample in the target domain, some relevant information from the source domain is available. In other words, the dataset should be structured as {(u_j, v_j, s_j)}_{j=1}^n, where s_j represents source domain information related to sample j (for instance, hardness or related information).
 
To apply our method effectively in your scenario, features x and u generally must belong to the same feature space (e.g., both x and u are alloy composition feature vectors). In this case, you would train a predictive model f(x) on the source dataset {(x, y)} and then generate the source-derived features s for the target data by applying f or extracting intermediate features from f. You would then form a dataset {(u, v, s)} to apply our method.
 
If x and u exist in different feature spaces (for example, x representing alloy composition and u representing structural features), our method cannot be directly applied. In such cases, additional steps like introducing an encoder to map u to the feature space of x would be required.
 
In short, for our method to be applicable, you need a target dataset of the form {(u_j, v_j, s_j)}_{j=1}^n, where each sample includes (target input, target output, and additional source-derived information).
 
Please let me know if anything needs further clarification.

**Comment by cyy521xx** [comment]
Hello, may I ask you a question? Regarding fine-tuning transfer in neural network training, I understand it this way: we should need a large amount of source domain data to train the source domain model, and then use target domain data to fine-tune the parameters of the source domain model. However, in the paper's experiments, the number of samples in the neural network's source domain model is the same as in the target domain. I'm not quite sure whether the neural network fine-tuning would perform better when there are many more source domain samples?


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "mshunya/AffineTL"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年4月27日(星期天) 下午5:26
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [mshunya/AffineTL] How should the program be modified when the feature distribution of the source domain is inconsistent with the source domain data distribution?" (Issue #1)





 
Closed #1 as completed.

**Comment by mshunya** [comment]
Thank you for your question.

> However, in the paper's experiments, the number of samples in the neural network's source domain model is the same as in the target domain.

In our experiments, the number of samples used for training the source model is actually larger than the number of samples used for fine-tuning on the target domain.
For example, in the `11_SARCOS` experiment, we trained the neural network source model with 30,000 samples, and then fine-tuned it on just 5–50 samples from the target task.
(Note: It appears that the code currently available in this GitHub repository is slightly outdated. Please refer to the Supplemental materials linked on our [paper's webpage](https://proceedings.neurips.cc/paper_files/paper/2023/hash/3819a070922cc0d19f3d66ce108f28e0-Abstract-Conference.html) to access the most recent code.)

> I'm not quite sure whether the neural network fine-tuning would perform better when there are many more source domain samples.

In general, it's theoretically and empirically suggested that transfer learning performance improves as the number of samples in the source domain increases. (For example, see [this paper](https://arxiv.org/abs/2108.11018).) We also expect this to be the case in our experiments.

Feel free to ask if you have further questions.

**Comment by cyy521xx** [comment]
Thank you very much for your reply. I will study your code carefully to better understand how to use AffineTL.

**Comment by cyy521xx** [comment]
I noticed that the program provided on GitHub normalizes the target variable y. Could you explain why normalization is applied to y? Thanks very much！

**Comment by cyy521xx** [comment]
I noticed that the program provided on GitHub normalizes the target variable y.&nbsp;Typically, only features are normalized. What are the benefits of normalizing labels?"&nbsp;Thanks very much！

**Comment by cyy521xx** [comment]
# Translation to English


Hello, may I ask you about an issue I'm encountering with Affine transfer learning? In my dataset (69 data points total), the Pearson correlation coefficient between source and target domains is 0.89. I'm using an 80:20 train-test split. I've found that directly using the source domain as one of the features with an Elastic Net model produces smaller gaps between training and test performance across 20 experiments compared to using Affine transfer learning (I only modified the input, keeping everything else consistent with your code). However, when the training set R² is lower than the test set R², the affine transfer performs very well. In my current situation, how should I adjust parameters to achieve better results using affine transfer learning? Thank you very much.

**Comment by cyy521xx** [comment]
Hello, may I ask you about an issue I'm encountering with Affine transfer learning? In my dataset (69 data points total), the Pearson correlation coefficient between source and target domains is 0.89. I'm using an 80:20 train-test split. I've found that directly using the source domain as one of the features with an Elastic Net model produces smaller gaps between training and test performance across 20 experiments compared to using Affine transfer learning (I only modified the input, keeping everything else consistent with your code). However, when the training set R² is lower than the test set R², the affine transfer performs very well. In my current situation, how should I adjust parameters to achieve better results using affine transfer learning? Thank you very much.

**Comment by mshunya** [comment]
Sorry, I'm not sure I fully understand your situation. Could you please confirm if the following points are correct?
- You have a dataset with 69 samples total, each consisting of inputs and outputs for both source and target domains, i.e., your data is {(xi, y_source,i, y_target,i)}_{i=1}^69 
- The Pearson correlation coefficient between y_source and y_target​ is 0.89.
- Predicting y_target​ using Elastic Net with inputs augmented by the source outputs (i.e., input x'=[x, y_source]) results in better performance than Affine transfer learning.
- Affine transfer learning tends to outperform Elastic Net specifically in cases where performance during training is not so good.

**Your question:** _How should I adjust the parameters to improve the overall performance of Affine Transfer?_

Note: If you could specify exactly which parts of my code you're using, I'll be able to provide a more accurate answer.

**Comment by cyy521xx** [comment]
I'm very glad to receive your reply.&nbsp;Yes, these four points are certain. I have also tried using multiple source domains. I conducted ten tests, and it seems that the performance of affine migration and the elastic net is similar.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "mshunya/AffineTL"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月12日(星期一) 上午9:41
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [mshunya/AffineTL] How should the program be modified when the feature distribution of the source domain is inconsistent with the source domain data distribution?" (Issue #1)



 mshunya left a comment (mshunya/AffineTL#1)
 
Sorry, I'm not sure I fully understand your situation. Could you please confirm if the following points are correct?
  
You have a dataset with 69 samples total, each consisting of inputs and outputs for both source and target domains, i.e., your data is {(xi, y_source,i, y_target,i)}_{i=1}^69
 
The Pearson correlation coefficient between y_source and y_target​ is 0.89.
 
Predicting y_target​ using Elastic Net with inputs augmented by the source outputs (i.e., input x'=[x, y_source]) results in better performance than Affine transfer learning.
 
Affine transfer learning tends to outperform Elastic Net specifically in cases where performance during training is not so good.
  
Your question: How should I adjust the parameters to improve the overall performance of Affine Transfer?
 
Note: If you could specify exactly which parts of my code you're using, I'll be able to provide a more accurate answer.

**Comment by mshunya** [comment]
In your case, since there's a strong linear correlation between the source and the target, it's natural that simply adding the source output as a feature in a regression model yields high accuracy.

When there's a strong linear correlation, our model's multiplicative term (specifically, the second term in the model formulation y=g1+g2*g3 ) tends to estimate g2 as a value close to 1, thereby reducing the benefit of our approach. By increasing the regularization parameter for the coefficient b (lambda2 in the code), our method can approximate the scenario of using x'=[x, y_source]  as input. However, since our code doesn't support L1 regularization like Elastic Net does, it would be challenging to learn sparse coefficients.

Unfortunately, apart from performing a more detailed hyperparameter grid search, there isn't much else I can recommend to improve performance. Even then, this approach is unlikely to yield significant improvements. It seems Elastic Net may be better suited for your dataset.

**Comment by cyy521xx** [comment]
Thank you for your explanation. I really appreciate it!




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "mshunya/AffineTL"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月12日(星期一) 中午1:36
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [mshunya/AffineTL] How should the program be modified when the feature distribution of the source domain is inconsistent with the source domain data distribution?" (Issue #1)



 mshunya left a comment (mshunya/AffineTL#1)
 
In your case, since there's a strong linear correlation between the source and the target, it's natural that simply adding the source output as a feature in a regression model yields high accuracy.
 
When there's a strong linear correlation, our model's multiplicative term (specifically, the second term in the model formulation y=g1+g2*g3 ) tends to estimate g2 as a value close to 1, thereby reducing the benefit of our approach. By increasing the regularization parameter for the coefficient b (lambda2 in the code), our method can approximate the scenario of using x'=[x, y_source]  as input. However, since our code doesn't support L1 regularization like Elastic Net does, it would be challenging to learn sparse coefficients.
 
Unfortunately, apart from performing a more detailed hyperparameter grid search, there isn't much else I can recommend to improve performance. Even then, this approach is unlikely to yield significant improvements. It seems Elastic Net may be better suited for your dataset.
